orchagent/intfsorch.cpp

```diff
@@ -250,26 +235,74 @@ bool IntfsOrch::addRouterIntfs(Port &port, sai_object_id_t virtual_router_id, Ma
         return false;
     }
 
-    m_portsOrch->setPort(port.m_alias, port);
+    gPortsOrch->setPort(port.m_alias, port);
 
     SWSS_LOG_NOTICE("Create router interface for port %s", port.m_alias.c_str());
 
     return true;
 }
 
-bool IntfsOrch::removeRouterIntfs(Port &port)
+bool IntfsOrch::removeRouterIntfs(string alias, IpPrefix ip_prefix)
 {
     SWSS_LOG_ENTER();
 
-    sai_status_t status = sai_router_intfs_api->remove_router_interface(port.m_rif_id);
+    if (m_syncdIntfses.find(alias) == m_syncdIntfses.end())
+        return true;
+
+    if (m_syncdIntfses[alias].ref_count > 2)
+    {
+        SWSS_LOG_NOTICE("Router interface is still referenced");
+        return false;
+    }
+
+    Port port;
+    assert(gPortsOrch->getPort(alias, port));
+
+    /* Remove subnet route */
+    sai_unicast_route_entry_t unicast_route_entry;
+    unicast_route_entry.vr_id = gVirtualRouterId;
+    unicast_route_entry.destination.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+    unicast_route_entry.destination.addr.ip4 = ip_prefix.getIp().getV4Addr() & ip_prefix.getMask().getV4Addr();
+    unicast_route_entry.destination.mask.ip4 = ip_prefix.getMask().getV4Addr();
+
+    sai_status_t status = sai_route_api->remove_route(&unicast_route_entry);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to remove subnet route pre:%s %d\n", ip_prefix.to_string().c_str(), status);
+        return false;
+    }
+
+    SWSS_LOG_NOTICE("Remove subnet route with prefix:%s", ip_prefix.to_string().c_str());
+    decreaseRouterIntfsRefCount(alias);
+
+    /* Remove action trap route */
+    unicast_route_entry.vr_id = gVirtualRouterId;
+    unicast_route_entry.destination.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+    unicast_route_entry.destination.addr.ip4 = ip_prefix.getIp().getV4Addr();
+    unicast_route_entry.destination.mask.ip4 = 0xFFFFFFFF;
+
+    status = sai_route_api->remove_route(&unicast_route_entry);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to remove action trap route ip:%s %d\n", ip_prefix.getIp().to_string().c_str(), status);
+        return false;
+    }
+
+    SWSS_LOG_NOTICE("Remove packet action trap route ip:%s\n", ip_prefix.getIp().to_string().c_str());
+    decreaseRouterIntfsRefCount(alias);
+    m_syncdIntfses[alias].ip_addresses.remove(ip_prefix.getIp());
```

Comment:
the counterpart, adding them is not in addRouterIntf. why deleting them should be in removeRouterIntf?